src/strings/ops.c

```diff
@@ -572,16 +572,77 @@ MVMint64 MVM_string_equal_at(MVMThreadContext *tc, MVMString *a, MVMString *b, M
     return MVM_string_substrings_equal_nocheck(tc, a, offset, bgraphs, b, 0);
 }
 
-MVMint64 MVM_string_equal_at_ignore_case(MVMThreadContext *tc, MVMString *a, MVMString *b, MVMint64 offset) {
-    MVMString *fca;
-    MVMString *fcb;
-    MVMROOT(tc, b, {
-        fca = MVM_string_fc(tc, a);
-        MVMROOT(tc, fca, {
-            fcb = MVM_string_fc(tc, b);
+MVMint64 MVM_string_equal_at_ignore_case(MVMThreadContext *tc, MVMString *haystack, MVMString *needle, MVMint64 h_offset) {
+    /* foldcase versions of needle */
+    MVMString *needle_fc;
+    MVMStringIndex h_graphs = MVM_string_graphs(tc, haystack);
+    MVMStringIndex n_graphs = MVM_string_graphs(tc, needle);
+    /* Difference in length before and after foldcase. Foldcase can sometimes
+     * increase the number of codepoints. */
+    /* Store the grapheme as we iterate through the haystack */
+    MVMGrapheme32 h_g, n_g;
+    MVMint64 i, j;
+    MVMint64 return_val = 1;
+    /* An additional needle offset which is used only when codepoints expand
+     * when casefolded. The offset is the number of additional codepoints that
+     * have been seen so haystack and needle stay aligned */
+    MVMint64 n_offset = 0;
+    fprintf(stderr, "starting function\n");
+    if (h_offset < 0) {
+        h_offset += h_graphs;
+        if (h_offset < 0)
+            h_offset = 0; /* XXX I think this is the right behavior here */
+    }
+    if (h_offset >= h_graphs) {
+        fprintf(stderr, "returning 0 because h_offset ≥ h_graphs\n");
+        return 0;
+    }
+    MVMROOT(tc, haystack, {
+        MVMROOT(tc, needle_fc, {
+            needle_fc = MVM_string_fc(tc, needle);
         });
     });
-    return MVM_string_equal_at(tc, fca, fcb, offset);
+
+    for (i = 0; i + h_offset < h_graphs && i + n_offset < n_graphs; i++) {
+        const MVMCodepoint *h_result_cps;
+        h_g = MVM_string_get_grapheme_at_nocheck(tc, haystack, h_offset + i);
+        MVMuint32 h_fc_cps = MVM_unicode_get_case_change(tc, h_g, MVM_unicode_case_change_type_fold, &h_result_cps);
+        fprintf(stderr, "h_fc_cps=%i first loop i=%li\n", h_fc_cps, i);
+        /* If we get 0 for the number that means the cp doesn't change when casefolded */
+        if (h_fc_cps == 0) {
+            n_g = MVM_string_get_grapheme_at_nocheck(tc, needle_fc, i + n_offset);
+            if (h_g != n_g) {
+                fprintf(stderr, "614 in first for loop h_g[%i] != n_g[%i] j=%li\n", h_g, n_g, j);
+                return_val = 0;
+                break;
+            }
+        }/*
```

Comment:
I think I'd go with less code, and just handle it in the path below. I doubt there'll be much of a performance difference; sure we avoid the branching associated with a loop, but only because we add a branch checking for that specific case. So it probably breaks even.